torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -140,87 +132,99 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)serializedPayload.length(),
+            (int64_t)work.message_.type()
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      if (work.message_.isShutdown()) {
+        std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+        pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
+      } else {
+        std::vector<torch::Tensor> payload = {
+            torch::from_blob(
+                (void *)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar}
+            )
+        };
+        std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+        pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
+        pg_->send(payload, work.to_, work.to_ /* channelTag */)->wait();
+      }
+    },
+    std::move(work)
+  ));
 }
 
-// making sure tensors are not deleted before send finishes
-void ProcessGroupAgent::sendLoop() {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+  threadPool_.run(std::bind(
+    [&](RecvWork& work) {
 
-  while (!stop_) {
-    if (sendQueue_.empty()) {
-      workProduceCV_.wait(lock);
-      continue;
-    }
+      torch::Tensor& payload = work.payload_;
+      std::stringstream ss(std::string(
+        (char*)payload.storage().data<signed char>(), payload.numel()));
 
-    auto work = std::move(sendQueue_.front());
-    sendQueue_.pop_front();
-    lock.unlock();
+      Message message = deserialize(work.type_, ss);
 
-    workConsumeCV_.notify_one();
-
-
-    std::stringstream ss;
-    serialize(work.message_, ss);
-    std::string str = ss.str();
-
-    std::vector<torch::Tensor> preamble = {
-      torch::tensor(
+      if (message.isRequest()) {
+        cb_(names_[work.from_], std::move(message), *this);
+      } else if (message.isResponse()) {
+        auto id = message.id();
         {
-          (int64_t)pg_->getRank(),
-          (int64_t)str.length(),
-        }, {torch::kLong})
-    };
-    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-    std::vector<torch::Tensor> payload =
-        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
-    pg_->send(payload, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-
-    lock.lock();
-  }
+          std::lock_guard<std::mutex> lock{futureMutex_};
+          futures_[id]->markCompleted(std::move(message));
+          futures_.erase(id);
+        }
+      } else {
+        // TODO: pass the error back to the caller instead of crashing here.
+        AT_ERROR("unrecognized message type ", message.type());
+      }
+    },
+    std::move(work)
+  ));
 }
 
 void ProcessGroupAgent::listenLoop() {
+  google::InitGoogleLogging("distributed.rpc");
```

Comment:
This line helped me get rid of the following warning for the `LOG(INFO)` (Thanks! @pritamdamania87 and @jamarshon):

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
```

But I don't feel right to call it here. I searched a bit, and saw `LOG` is used extensively in `caffe2/` but almost not used at all in `torch/` or `aten/`. @soumith @dzhulgakov @gchanan @ezyang In general, what are the recommended API for logging? Or are we intentionally avoiding that?

Reply:
I think @dzhulgakov knows what the current state here is